torch/_prims/__init__.py

```diff
@@ -2588,3 +2594,132 @@ def _uniform_aten(
     impl_aten=_uniform_aten,
     doc=_uniform_doc,
 )
+
+
+def _fft_r2c_meta(
+    input: TensorLike,
+    *,
+    dim: DimsSequenceType,
+    onesided: bool,
+) -> TensorLikeType:
+    dim = utils.canonicalize_dims(input.ndim, dim)
+    utils.validate_no_repeating_dims(dim)
+
+    shape = list(input.shape)
+    if onesided:
+        last_dim = dim[-1]
+        shape[last_dim] = shape[last_dim] // 2 + 1
+
+    dtype = utils.corresponding_complex_dtype(input.dtype)
+    # TODO: Actual result strides may differ, is that okay?
+    strides = utils.make_contiguous_strides_for(shape)
+    return TensorMeta(shape=shape, strides=strides, dtype=dtype, device=input.device)
+
+
+def _fft_r2c_aten(
+    input: TensorLike,
+    *,
+    dim: DimsSequenceType,
+    onesided: bool,
+):
+    normalization = 0  # No normalization
+    dim = utils.canonicalize_dims(input.ndim, dim)
```

Comment:
dim canonicalization usually happens in references, not primitives, so primitives don't have to concern themselves with these details

in this case a wrapper `refs.fft_r2c` can be added to handle input processing and error checking